### PR TITLE
Improvements to Navigation and Localization Interfaces

### PR DESCRIPTION
**Pull request by randaz81**
Navigation_Improvements {#devel}
----------------

### libYARP_dev

* Added the following methods to `yarp::dev::ILocalization2D` interface
  * `startLocalizationService();`
  * `stopLocalizationService();`
  * `getCurrentPosition(yarp::dev::Nav2D::Map2DLocation& loc, yarp::sig::Matrix& cov);`
  * `setInitialPose(const yarp::dev::Nav2D::Map2DLocation& loc, const yarp::sig::Matrix& cov);`

* added data type `Vec2D<size_t>`

* `yarp::dev::Nav2D::Map2DPath` is now replacing std::vector<yarp::dev::Nav2D::Map2DLocation> in all APIs

* Added new class `yarp:dev::Nav2D::MapGrid2DInfo` which contains infos about the name, the resolution and the origin of
the map (e.g. `yarp:dev::Nav2D::MapGrid2DOrigin`) etc. The new class also contains several utility functions which were
previously belonging to `yarp:dev::Nav2D::MapGrid2D`. These function have been moved into `yarp:dev::Nav2D::MapGrid2DInfo`. They are:
  * `XYWorld cell2World(XYCell cell) const;`
  * `yarp::dev::Nav2D::Map2DLocation toLocation(XYCell cell) const;`
  * `XYCell toXYCell(yarp::dev::Nav2D::Map2DLocation loc) const;`
  * `XYCell world2Cell(XYWorld world) const;`
  * `yarp::dev::Nav2D::Map2DLocation toLocation(XYWorld cell) const;`
  * `XYWorld toXYWorld(yarp::dev::Nav2D::Map2DLocation loc) const;`

* Created namespace `yarp::dev::Nav2D` and moved into it the following classes:
  * `Map2DArea`
  * `Map2DPath`
  * `Map2DLocation`
  * `MapGrid2D`
  * `MapGrid2DInfo`

* The class `yarp::Nav2D::Map2DLocation` is now generated by Thrift
* The class `yarp::Nav2D::Map2DPath` is now generated by Thrift

* Added new path-related methods to `yarp::dev::IMap2D` interface:
  * `virtual bool storePath(std::string path_name, yarp::dev::Nav2D::Map2DPath path) = 0;`
  * `virtual bool getPath(std::string path_name, yarp::dev::Nav2D::Map2DPath& path) = 0;`
  * `virtual bool getPathsList(std::vector<std::string>& paths) = 0;`
  * `virtual bool deletePath(std::string path_name) = 0;`
  * `virtual bool renamePath(std::string original_name, std::string new_name) = 0;`
  * `virtual bool clearAllPaths() = 0;`

### CODE REVIEW

File: src/libYARP_dev/src/yarp/dev/Map2DPath.cpp [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+            this->waypoints[i].x != r.waypoints[i].x ||
+            this->waypoints[i].y != r.waypoints[i].y ||
+            this->waypoints[i].theta != r.waypoints[i].theta) return false;
+    return true;
+}
+
+std::string Map2DPath::toString() const
+{
+    std::ostringstream stringStream;
+    stringStream.precision(-1);
```
Comment (drdanz) [inline src/libYARP_dev/src/yarp/dev/Map2DPath.cpp]: Codacy Issue found: [Invalid stringStream.precision() argument nr 1. The value is -1 but the valid values are '0:'.](https://app.codacy.com/app/robotology/yarp/pullRequest?prid=4615567)

Comment (drdanz) [inline src/libYARP_dev/src/yarp/dev/Map2DPath.cpp]: Codacy Issue found: [Invalid stringStream.precision() argument nr 1. The value is -1 but the valid values are '0:'.](https://app.codacy.com/app/robotology/yarp/pullRequest?prid=4615567)

Comment (drdanz) [inline src/libYARP_dev/src/yarp/dev/Map2DPath.cpp]: Codacy Issue found: [Invalid stringStream.precision() argument nr 1. The value is -1 but the valid values are '0:'.](https://app.codacy.com/app/robotology/yarp/pullRequest?prid=4615567)

Comment (drdanz) [inline src/libYARP_dev/src/yarp/dev/Map2DPath.cpp]: Codacy Issue found: [Invalid stringStream.precision() argument nr 1. The value is -1 but the valid values are '0:'.](https://app.codacy.com/app/robotology/yarp/pullRequest?prid=4615567)

File: src/libYARP_dev/src/CMakeLists.txt [hunk 2]
```diff
@@ -200,18 +203,35 @@ if(TARGET YARP::YARP_math)
...
+else()
+yarp_idl_to_dir(INPUT_FILES
+                idl/stateExt.thrift
+                idl/audioBufferSizeData.thrift 
                 OUTPUT_DIR "${CMAKE_CURRENT_SOURCE_DIR}/idl_generated_code"
                 SOURCES_VAR YARP_dev_idl_SRCS
                 HEADERS_VAR YARP_dev_idl_all_HDRS
                 INCLUDE_DIRS_VAR YARP_dev_idl_INCLUDE_DIRS
                 PLACEMENT MERGED)
+endif()
```
Comment (drdanz) [inline src/libYARP_dev/src/CMakeLists.txt]:
```suggestion
set(YARP_dev_IDL idl/stateExt.thrift
                 idl/audioBufferSizeData.thrift)

if(TARGET YARP::YARP_math)
  list(APPEND YARP_dev_IDL idl/stateExt.thrift
                           idl/audioBufferSizeData.thrift 
                           idl/Map2DLocationData.thrift
                           idl/Map2DPathData.thrift)
endif()

yarp_idl_to_dir(INPUT_FILES ${YARP_dev_IDL}
                OUTPUT_DIR "${CMAKE_CURRENT_SOURCE_DIR}/idl_generated_code"
                SOURCES_VAR YARP_dev_idl_SRCS
                HEADERS_VAR YARP_dev_idl_all_HDRS
                INCLUDE_DIRS_VAR YARP_dev_idl_INCLUDE_DIRS
                PLACEMENT MERGED)
```

Comment (randaz81) [reply src/libYARP_dev/src/CMakeLists.txt]: fixed as suggested

**Pull request merged**